src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -719,16 +801,38 @@ private static Array BindArray(Type type, IEnumerable? source, IConfiguration co
             return result;
         }
 
+        private static bool TypeIsADictionaryInterface(Type type)
+        {
+            if (!type.IsInterface || !type.IsConstructedGenericType) { return false; }
+
+            Type genericTypeDefinition = type.GetGenericTypeDefinition();
+            return genericTypeDefinition == typeof(IDictionary<,>)
+                || genericTypeDefinition == typeof(IReadOnlyDictionary<,>);
+        }
+
         private static bool IsArrayCompatibleReadOnlyInterface(Type type)
         {
             if (!type.IsInterface || !type.IsConstructedGenericType) { return false; }
 
             Type genericTypeDefinition = type.GetGenericTypeDefinition();
             return genericTypeDefinition == typeof(IEnumerable<>)
+                || genericTypeDefinition == typeof(ICollection<>)
```

Comment:
I don't think this fits in with the name of the method: `IsArrayCompatibleReadOnlyInterface` - `ICollection<>` is not "read-only".

Reply:
I think the method is answering the question 'Should the type be treated as immutable?" - e.g.  if it's one of these types, then return `true` to say **not** to append to the source, but clone it to a new collection.
I've renamed it as such, although I do think it could be clearer. Do you have any thoughts?

Reply:
Why do we need to check for `ICollection<>` in both places?

Reply:
I think the second check for `ICollection<>` is looking for an `Add` method to determine whether to set (add) members of the collection, or whether to bind to properties (line 358).

Do you have any thoughts on how this could be made clearer? To be perfectly honest, I do find this method rather confusing!